### DISCUSSION TITLE
Fixture for specific module, class, package with autouse option

### DISCUSSION THREAD

**Discussion by AliyevH** [pytest-dev/pytest#10280] [General]
Hi.
Suppose we have such fixture in conftest.py globally.
But we want it to run with scope session, but for specific classes or modules.
Without creating dirs, or writing it in specific py file, can we say it somehow to run only in specified places?

```
import somepackage
@pytest.fixture(scope='session', autouse=True)
def some_mock():
    with mock.patch('somepackage.somemethod') as _fixture:
        yield _fixture
```
For example
```
import somepackage
@pytest.fixture(scope='session', autouse=True, class='SomeClass', package='somepackage.py')
def some_mock():
    with mock.patch('somepackage.somemethod') as _fixture:
        yield _fixture
```

**Comment by nicoddemus** [comment]
Hi,

> But we want it to run with scope session, but for specific classes or modules.

I assume you mean test modules and test classes right? Otherwise your question does not make sense to me, because pytest runs tests, not packages/classes.

If you do mean test modules/classes, one way would be to use [`request`](https://docs.pytest.org/en/stable/reference/reference.html#request) to see if the test requesting module is the one your fixture should run:

```python
import somepackage_test

@pytest.fixture(scope='session', autouse=True)
def some_mock(request):
    if request.module is not somepackage_test:
        return
    with mock.patch('somepackage.somemethod') as _fixture:
        yield _fixture
```

Same logic for [`request.cls`](https://docs.pytest.org/en/stable/reference/reference.html#pytest.FixtureRequest.cls).

**Comment by sparrowt** [comment]
An alternative is to remove `autouse=True` and then to bulk apply it to all tests in certain test files by adding this at top module level:
```python
pytestmark = pytest.mark.usefixtures("your_fixture_name")
```
this saves you adding `your_fixture_name` as an argument of every test in that file.

**Reply by clemensfrehsner** [reply]
this is the first time i've ever commented on something. you're a saint, that's exactly what i needed
